kusto/tokenprovider.go

```diff
@@ -0,0 +1,59 @@
+package kusto
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+type TokenProvider struct {
+	tokenCred   azcore.TokenCredential //Holds the received token credential as per the authorization
+	tokenScheme string                 //Contains token scheme for tokenprovider
+	customToken string                 //Holds the custom auth token to be used for authorization
+	initOnce    *sync.Once             //To ensure tokenprovider will be initialized only once while aquiring token
+	init        func()                 //Initialiser for initialising the tokencredential and cloudinfo
+	ci          CloudInfo              //Contains cloud setting meto information
+	scopes      []string               //Contains scopes of the auth token
+	err         error                  //To monitor errors in the process of init tokenprovider
+}
+
+// tokenProvider need to be received as reference, to reflect updations to the structs
+func (tkp *TokenProvider) AcquireToken(ctx context.Context) (string, string, error) {
+	if !isEmpty(tkp.customToken) {
+		return tkp.customToken, tkp.tokenScheme, nil
+	}
+
+	if tkp.tokenCred == nil {
```

Comment:
Shouldn't the check be on "initOnce"?

Reply:
@AsafMah  The sync.Once should not be used as a communication to check if it is already done the task. An initializer flag can be used to do that, but that we avoided by checking the tokenCred object itself.  Are we missing any cases if we use tokenCred as a check?